pkg/securitycontext/admission/admission.go

```diff
@@ -0,0 +1,177 @@
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/auth/user"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints"
+)
+
+func init() {
+	kadmission.RegisterPlugin("SecurityContextConstraint", func(client client.Interface, config io.Reader) (kadmission.Interface, error) {
+		return NewConstraint(client), nil
+	})
+}
+
+type constraint struct {
+	*kadmission.Handler
+	client client.Interface
+}
+
+var _ kadmission.Interface = &constaint{}
+
+func NewConstraint(client client.Interface) kadmission.Interface {
+	return &constraint{
+		Handler: kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		client:  client,
+	}
+}
+
+func (c *constraint) Admit(a kadmission.Attributes) error {
+	// 1.  find all SCCs the pod has access to:
+	// 2.  Fully resolve each SCC
+	//     1.  determine uid from the SCC's run as user policy
+	//     2.  set the uid on the SCC
+	// 3.  Match the pod's SC to an SCC; for each SCC:
+	//     1.  generate the SC
+	//     2.  validate against the pod's SC
+
+	if a.GetResource() != "pods" {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(kapi.Pod)
+	if !ok {
+		return errors.NewBadRequest("a pod was received, but could not convert the request object.")
+	}
+
+	ns := a.GetNamespace()
+	matchedConstraints, err := getMatchingSecurityContextConstraints(c.client, ns, pod)
+	if err != nil {
+		return err
+	}
+
+	for _, constraint := range matchedConstraints {
+		if constraint.RunAsUser.Type == kapi.RunAsUserStrategyMustRunAs && constraint.RunAsUser.UID == nil {
+			uid, err := getPreallocatedUID(constraint)
+			if err != nil {
+				return err
+			}
+
+			constraint.RunAsUser.UID = &uid
+		}
+
+		if constraint.SELinuxContext.Type == kapi.SELinuxStrategyMustRunAs && constraint.SELinuxContext.SELinuxOptions.Level == "" {
+			level, err := getPreallocatedLevel(constraint)
+			if err != nil {
+				return err
+			}
+
+			constraint.SELinuxContext.SELinuxOptions.Level = level
+		}
+	}
+
+	providers := make([]SecurityContextConstraintsProvider)
+
+	for _, constraint := range matchedConstraints {
+		provider, err := securitycontextconstraints.NewSimpleProvider(constraint, c.client)
+		if err != nil {
+			return err
+		}
+
+		providers = append(providers, provider)
+	}
+
+Containers:
+	for i, container := range pod.Spec.Containers {
+		for _, provider := range providers {
+			// Create a security context for this provider
+			context, err := provider.CreateSecurityContext(pod, container, nil)
+
+			// Create a deep copy of the pod to test with the security context
+			podCopyObj, err := kapi.Scheme.Copy(pod)
+			if err != nil {
+				return err
+			}
+			podCopy, ok := podCopyObj.(kapi.Pod)
```

Comment:
I've been thinking about this copy since it doesn't sit right with me.  Perhaps we should change the interface to pass in the context that will be used and avoid the copy altogether.  The provider and strats should know to validated the passed in context and not what is currently on the container and it is up to admission, as done below, to set the context if it validates.